Robot

amymhaddad opened the pull request:
@paulghaddad Please review the attached program and tests. 

I have one question: I was trying to use a fixture on this test series (to use on at least two of the tests). Do I have this syntax correct:

```
import pytest

@pytest.fixture
def robot():
     return Robot()
```

Then, in my test I should be able to call the object:

```
def test_reset_name_to_different_name():
    assert robot.reset() != robot.name
```

However, the above test doesn't work, and I'm not sure why. The fixture object should create a new instance of the Robot class, right? I get an error saying the reset() function is not found.

amymhaddad committed:
Adding robot name files; got algorithm to generate a random name working

amymhaddad committed:
Algorithm works, need to update tests

amymhaddad committed:
Final changes to robot name for PR review

paulghaddad commented:
@amymhaddad 

You need to pass the fixture in as an argument to the test:

```
def test_reset_name_to_different_name(robot):
    assert robot.reset() != robot.name
```

Note how the smtp_connection fixture is passed in here:

https://docs.pytest.org/en/latest/fixture.html#fixtures-as-function-arguments

paulghaddad commented on robot_name/robot_name.py:
```diff
@@ -0,0 +1,14 @@
+    def __init__(self):
+        self.name = self.create_random_name()
+
+    def create_random_name(self):
```
Nice job here! 💯

paulghaddad commented on robot_name/test_robot_name.py:
```diff
@@ -0,0 +1,17 @@
+
+def test_reset_name_to_different_name():
+    robot = Robot()
+    assert robot.reset() != robot.name
```
I think a better way to test this is:

```
def test_reset_name_to_different_name(robot):
     original_name = robot.name # this is the setup; you're saving the original name

     robot.reset() # exercise the method

     assert original_name != robot.name # verify the new name is different
```

paulghaddad commented on robot_name/test_robot_name.py:
```diff
@@ -0,0 +1,17 @@
+
+def test_length_robot_name():
+    robot = Robot()
+    assert len(robot.name) == 5
```
a better way to test this is a regex match :)

paulghaddad commented on robot_name/robot_name.py:
```diff
@@ -0,0 +1,14 @@
+from string import digits as digit
+
+
+class Robot:
```
Everything in this class is spot-on!

paulghaddad reviewed (changes requested):
Really nice work! You're doing great on testing. Just a couple more advanced pointers. 🌵

amymhaddad committed:
WIP - updating tests from PR review feedback

amymhaddad committed:
Rewrote tests 1 and 3 and both work

amymhaddad commented:
@paulghaddad Great comments!

Here are my changes in `test_robot_name.py`:
1. Added fixture object and used fixture object in ALL three tests
```
@pytest.fixture
def robot():
    return Robot()
```

2. Updated ALL tests. 
I was getting an error when I tried your suggestion for testing reset() (in test 2). So I re-wrote the test as:

```
def test_reset_name_to_different_name(robot):
    original_name = robot.name
    assert original_name != robot.reset()
```

I used regex for the last test:

```
def test_robot_name_contains_correct_values(robot):
    result = re.match(r"\D{2}\d{3}", robot.name)
    assert result.group() == robot.name
```

paulghaddad commented on robot_name/test_robot_name.py:
```diff
@@ -0,0 +1,23 @@
+
+def test_reset_name_to_different_name(robot):
+    original_name = robot.name
+    assert original_name != robot.reset()
```
This makes sense.

paulghaddad commented on robot_name/test_robot_name.py:
```diff
@@ -0,0 +1,23 @@
+
+
+def test_robot_name_contains_correct_values(robot):
+    result = re.match(r"\D{2}\d{3}", robot.name)
```
Excellent use of a regex. And I didn't know about `.group()`.

paulghaddad reviewed (approved):
This is spot on. Excellent. 💯
